### PR TITLE
Support asa ID in `deployer.OptIn*`

### PR DESCRIPTION
**Pull request by man-of-code**

### CODE REVIEW

**Comment by amityadav0** [comment]
Please add a test, showing we can retreive asset info from checkpoint using asa id. You can refer to some Tests in checkpoints test section.

**Comment by man-of-code** [comment]
> Please add a test, showing we can retreive asset info from checkpoint using asa id. You can refer to some Tests in checkpoints test section.

@amityadav0  My understanding is that this task was to provide support for ASA not deployed by `deployer` (not existing in checkpoints). @ratik21 Has added a function in this [PR](https://github.com/scale-it/algo-builder/pull/293) with the same functionality you are talking about.

**Comment by amityadav0** [comment]
@man-of-code right, we may have overestimated this task.

File: packages/algob/src/internal/deployer.ts [hunk 1]
```diff
@@ -184,27 +184,51 @@ class DeployerBasicMode {
...
    * @param flags Transaction flags
    */
-  async optInLsigToASA (asaName: string, lsig: LogicSig, flags: rtypes.TxParams): Promise<void> {
-    await this.algoOp.optInLsigToASA(asaName, this._getASAInfo(asaName).assetIndex, lsig, flags);
+  async optInLsigToASA (asa: string, lsig: LogicSig, flags: rtypes.TxParams): Promise<void> {
+    try {
+      const asaId = this._getASAInfo(asa).assetIndex;
+      await this.algoOp.optInLsigToASA(asa, asaId, lsig, flags);
+    } catch (error) {
+      console.log(error);
```
Comment (amityadav0) [inline packages/algob/src/internal/deployer.ts:225]: let's change this log here, maybe to asset not found in checkpoints proceeding to process it as asset ID

File: packages/algob/src/internal/deployer.ts [hunk 2]
```diff
@@ -184,27 +184,51 @@ class DeployerBasicMode {
...
+      if (!Number(asa)) {
+        throw Error("Please provide a valid Number to be used as ASA ID");
```
Comment (amityadav0) [inline packages/algob/src/internal/deployer.ts:227]: This error should show both cases - 1. asset not in checkpoints also asset is not valid ASA ID

File: packages/algob/src/internal/deployer.ts [hunk 3]
```diff
@@ -184,27 +184,51 @@ class DeployerBasicMode {
...
+        this._getAccount(accountName),
+        flags);
+    }
   }
 
   /**
    * Description: Opt-In to ASA for a contract account (represented by logic signture).
    * The opt-in transaction is signed by the logic signature
-   * @param asaName ASA name
+   * @param asa ASA (name/ID)
```
Comment (amityadav0) [inline packages/algob/src/internal/deployer.ts:216]: Please add here ID can be used for ASA not existing in checkpoints

File: packages/algob/src/internal/deployer.ts [hunk 4]
```diff
@@ -184,27 +184,51 @@ class DeployerBasicMode {
...
+  async optInAcountToASA (asa: string, accountName: string, flags: rtypes.TxParams): Promise<void> {
+    try {
+      const asaId = this._getASAInfo(asa).assetIndex;
+      await this.algoOp.optInAcountToASA(
+        asa,
+        asaId,
+        this._getAccount(accountName),
+        flags);
+    } catch (error) {
+      console.log(error);
```
Comment (amityadav0) [inline packages/algob/src/internal/deployer.ts:200]: Please update this function as per suggestions in Lsig function

**Review by amityadav0 — CHANGES_REQUESTED** [review]
I don't think we support this in `executeTransaction`. Please add these checks there.

**Comment by man-of-code** [comment]
> I don't think we support this in `executeTransaction`. Please add these checks there.

@amityadav0 while executing any asset related transaction using `executeTransaction` we are already sending AssetID.

**Review by amityadav0 — APPROVED** [review]

**Pull request merged**